File: MarkdownEditor/MarkdownEditor/Sources/Converter/MarkdownRendererConverter.m
//
//  MarkdownRendererConverter.m
//  MarkdownEditor
//
//  Created by Iwaki Satoshi on 2018/02/27.
//  Copyright © 2018 Satoshi Iwaki and 2026 Vincent Janelle. All rights reserved.
//

#import "MarkdownRendererConverter.h"
#import "Logger.h"
#import "MarkdownEditorLite-Swift.h"

@implementation MarkdownRendererConverter {
    NSString *_html;
    NSString *_title;
    NSString *_format;
    NSString *_header;
    NSString *_css;
}

- (instancetype)initWithTitle:(NSString *)title
                       format:(NSString *)format
                       header:(nullable NSString *)header
                          css:(nullable NSString *)css {
    self = [super initWithTitle:title];
    if (self) {
        _format = [format copy];
        _header = [header copy];
        _css = [css copy];
        _html = @"<html><body></body></html>";
    }
    return self;
}

- (NSString *)html {
    return _html;
}

- (NSString *)formattedStringWithString:(NSString *)string format:(TextConverterFormat)format {
    switch (format) {
        case TextConverterFormatBold:
            return [NSString stringWithFormat:@"**%@**", string];
        case TextConverterFormatItalic:
            return [NSString stringWithFormat:@"*%@*", string];
        case TextConverterFormatStrikeThrough:
            return [NSString stringWithFormat:@"~~%@~~", string];
        case TextConverterFormatCode: {
            return [NSString stringWithFormat:@"```\n%@\n```", string];
        }
        case TextConverterFormatLink: {
            return [NSString stringWithFormat:@"[%@](url)", string];
        }
        case TextConverterFormatQuote: {
            NSArray<NSString *> *lines = [string componentsSeparatedByString:@"\n"];
            NSMutableString *formattedString = [@"" mutableCopy];
            for (NSString *line in lines) {
                [formattedString appendString:@"> "];
                [formattedString appendString:line];
                [formattedString appendString:@"\n"];
            }
            return formattedString;
        }
        case TextConverterFormatListBulleted: {
            NSArray<NSString *> *lines = [string componentsSeparatedByString:@"\n"];
            NSMutableString *formattedString = [@"" mutableCopy];
            for (NSString *line in lines) {
                [formattedString appendString:@"- "];
                [formattedString appendString:line];
                [formattedString appendString:@"\n"];
            }
            return formattedString;
        }
        case TextConverterFormatListNumbered: {
            NSArray<NSString *> *lines = [string componentsSeparatedByString:@"\n"];
            NSMutableString *formattedString = [@"" mutableCopy];
            for (NSString *line in lines) {
                [formattedString appendString:@"1. "];
                [formattedString appendString:line];
                [formattedString appendString:@"\n"];
            }
            return formattedString;
        }
        default:
            break;
    }
    return string;
}

- (void)setContentWithString:(NSString *)string {
    @synchronized (self) {
        NSString *headerHTML = nil;
        if (_header) {
            NSString *path = [NSBundle.mainBundle pathForResource:_header ofType:@""];
            headerHTML = [NSString stringWithContentsOfFile:path encoding:NSUTF8StringEncoding error:nil];
        }

        MarkdownHTMLRenderer *renderer = [[MarkdownHTMLRenderer alloc] init];
        _html = [renderer htmlStringFromMarkdown:string css:self.css header:headerHTML];

        LogD(@"*** HTML ***");
        LogD(@"%@", _html);
    }
}

@end

File: MarkdownEditor/MarkdownEditor/Sources/Converter/MarkdownRendererConverter.swift
import Foundation

@objc(MarkdownRendererConverter)
class MarkdownRendererConverter: TextConverter {
    @objc let header: String?

    private let rendererFormat: String
    private let rendererCSS: String?
    private var renderedHTML = "<html><body></body></html>"

    @objc(initWithTitle:format:header:css:)
    init(title: String, format: String, header: String?, css: String?) {
        self.rendererFormat = format
        self.header = header
        self.rendererCSS = css
        super.init(title: title)
    }

    @objc override var format: String {
        rendererFormat
    }

    @objc override var css: String? {
        rendererCSS
    }

    @objc override var html: String {
        renderedHTML
    }

    @objc override func formattedString(with string: String, format: TextConverterFormat) -> String {
        Self.markdownFormattedString(string, format: format)
    }

    @objc override func setContent(with string: String) {
        objc_sync_enter(self)
        defer { objc_sync_exit(self) }

        let headerHTML: String?
        if let header,
           let path = Bundle.main.path(forResource: header, ofType: "") {
            headerHTML = try? String(contentsOfFile: path, encoding: .utf8)
        } else {
            headerHTML = nil
        }

        let renderer = MarkdownHTMLRenderer()
        renderedHTML = renderer.htmlString(fromMarkdown: string, css: rendererCSS, header: headerHTML)
        logDebug("*** HTML ***")
        logDebug(renderedHTML)
    }

    static func markdownFormattedString(_ string: String, format: TextConverterFormat) -> String {
        switch format {
        case .bold:
            return "**\(string)**"
        case .italic:
            return "*\(string)*"
        case .strikeThrough:
            return "~~\(string)~~"
        case .code:
            return "```\n\(string)\n```"
        case .link:
            return "[\(string)](url)"
        case .quote:
            return string.components(separatedBy: "\n").map { "> \($0)" }.joined(separator: "\n") + "\n"
        case .listBulleted:
            return string.components(separatedBy: "\n").map { "- \($0)" }.joined(separator: "\n") + "\n"
        case .listNumbered:
            return string.components(separatedBy: "\n").map { "1. \($0)" }.joined(separator: "\n") + "\n"
        }
    }
}

@objc(MarkdownConverter)
class MarkdownConverter: MarkdownRendererConverter {
    override init(title: String, format: String, header: String?, css: String?) {
        super.init(title: title, format: format, header: header, css: css)
    }

    @objc init() {
        super.init(title: "Markdown", format: "markdown", header: "markdown-header.txt", css: "markdown.css")
    }
}

@objc(GfmConverter)
class GfmConverter: MarkdownConverter {
    @objc override init() {
        super.init(title: "GitHub Flavored Markdown", format: "gfm", header: "gfm-header.txt", css: "gfm.css")
    }
}

@objc(StrictMarkdownConverter)
class StrictMarkdownConverter: MarkdownConverter {
    @objc override init() {
        super.init(title: "Strict Markdown", format: "markdown_strict", header: nil, css: "markdown.css")
    }
}
